### ISSUE TITLE
Emulation pauses on Windows while window is being moved

### ISSUE THREAD

**Issue by weirddan455** [dosbox-staging/dosbox-staging#3554]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Unsure if this is expected behavior but it doesn't not occur on Linux.

### Steps to reproduce the behaviour.

Explain how to reproduce
1.  Start a game
2.  Move the window
3.  Everything pauses until the window stops moving

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by Grounded0** [comment]
Doesn't happen on macOS either.

**Comment by johnnovak** [comment]
Yeah, it's "standard" on Windows @weirddan455. I think GLFW behaves the same way. Let's call improving this an enhancement instead.

You need to jump through quite a few hoops to keep the app running while moving the window and resizing. Although resizing is definitely harder.

If I'm not mistaken, it involves moving all rendering and blocking off the main thread, so the main thread can continue to process windowing system events without blocking. This has a lot to do with the windowing system event handling pecularities per OS, so I think that's why SDL does not want to normalise that. Too app-specific and "opinionated", and it result in a lot of complexity vs keeping things on a main thread (it effectively forces you into a multithreaded app design) 

Maybe best to defer this to after the SDL3 migration? Our SDL and init code is messy already.

**Comment by johnnovak** [comment]
I've done some digging around @weirddan455 and this is our lucky day: this is supposed to be fixed in SDL 2.30.0+ 🥳 
Just we're on 2.26.5 on Windows in 0.81.x because we had to revert due to some showstopper regressions.

See relevant thread and the very last comment:
https://github.com/libsdl-org/SDL/issues/1059

So we should get this for free in our next release with the SDL upgrade.

**Comment by johnnovak** [comment]
Well, just tested Staging with 2.30.0 on Windows and upgrading SDL is not enough; the emulation is still being paused when you drag the window by the titlebar and move it around... 😐

**Comment by weirddan455** [comment]
It's probably not viable to run emulation ticks inside a `SDL_EVENT_WINDOW_EXPOSED` handler.  Usually all you would want to do there is blit the current framebuffer to the screen.  It's meant for applications that sit idle most of the time to be able to re-draw themselves on command (for example when another window moves over the top of it).

**Comment by johnnovak** [comment]
> It's probably not viable to run emulation ticks inside a `SDL_EVENT_WINDOW_EXPOSED` handler. Usually all you would want to do there is blit the current framebuffer to the screen. It's meant for applications that sit idle most of the time to be able to re-draw themselves on command (for example when another window moves over the top of it).

Yeah, exactly my thoughts. This solution could only work if the whole emulation is moved off the main thread.

**Comment by joncampbell123** [comment]
DOSBox-X developer here: The way Windows works when DefWindowProc() sees mouse input messages related to moving or sizing, is that it goes into it's own message loop to handle it, blocking _your_ process.

Rather than trying to run the emulation loop in a second thread, DOSBox-X takes the approach of letting the main thread manage a child window within the main window, and a second thread takes care of the outer main window. Thus, the second thread is blocked by DefWindowProc when you move the window while DOSBox-X can continue to run uninterrupted.

The same blockage occurs when DefWindowProc() is handling user input to the menu bar.

This was a modification of the SDL1 library made quite a while ago and it's been part of DOSBox-X since.

I would be OK with it of SDL2 were to adapt some form of that for other projects including Staging.

**Comment by johnnovak** [comment]
Super informative comment @joncampbell123 as always, thanks for chiming in. I do hope SDL will implement an OS-agnostic way to achieve this so we don't have to deal with all this low-level mucking around... Maybe in SDL3, one day 🤞🏻

**Comment by weirddan455** [comment]
I just happened to across an article in the SDL3 wiki discussing this issue. It has a potential solution.

https://wiki.libsdl.org/SDL3/AppFreezeDuringDrag

**Comment by johnnovak** [comment]
> I just happened to across an article in the SDL3 wiki discussing this issue. It has a potential solution.
> 
> https://wiki.libsdl.org/SDL3/AppFreezeDuringDrag

We could use this to still present frames at a potentially lower rate (whatever the callback rate is) during resize. So graphics output won't completely freeze, but probably will become a bit choppy.

We still need to move the entire emulation off the main thread, and only run the SDL event loop in the main thread, otherwise audio and CPU emulation will be still completely stalled by resizing the window.

For example, play a game that ouputs MIDI music and start resizing. The MIDI notes will ring out while resizing, meaning the audio callback works fine during resize (expected), and the emulated MIDI device keeps feeding the audio buffers as it runs in a separate threads, _but_ it gets no new notes because the MIDI notes are fed by the CPU emulation in the main thread (now stalled).

So this would be just part of the puzzle; moving the entire emulation code off the main thread still needs to happen. If the CPU emulation is stalled, this callback won't help us; it needs to keep going like clockwork.
